File: sdks/typescript/src/guard.ts
/** Local Policy Enforcement Points: ToolGuard + MCPGuard. */

import { PermissionDeniedError } from "./errors.js";
import { AuthzClient } from "./client.js";

export class ToolGuard {
  private readonly permissions: Set<string>;
  constructor(permissions: Iterable<string>) {
    this.permissions = new Set(permissions);
  }
  isAllowed(resource: string, action: string): boolean {
    return this.permissions.has(`${resource}.${action}`);
  }
  require(resource: string, action: string): void {
    if (!this.isAllowed(resource, action)) {
      throw new PermissionDeniedError(`${resource}.${action}`);
    }
  }
}

export class MCPGuard {
  private readonly permissions: Set<string>;
  constructor(permissions: Iterable<string>) {
    this.permissions = new Set(permissions);
  }
  isAllowed(server: string, action: string): boolean {
    return this.permissions.has(`mcp.${server}.${action}`);
  }
  require(server: string, action: string): void {
    if (!this.isAllowed(server, action)) {
      throw new PermissionDeniedError(`mcp.${server}.${action}`);
    }
  }
  allowedTools(server: string): string[] {
    const prefix = `mcp.${server}.`;
    return [...this.permissions]
      .filter((p) => p.startsWith(prefix))
      .map((p) => p.slice(prefix.length));
  }
}

/** Bundle: preload effective permissions and return guards bound to that snapshot. */
export async function startAgentSession(
  client: AuthzClient,
  opts: {
    tenantId: string;
    applicationId: string;
    userId: string;
    agentId: string;
  },
): Promise<{ tools: ToolGuard; mcp: MCPGuard; permissions: Set<string> }> {
  const permissions = await client.getEffectivePermissions({
    tenantId: opts.tenantId,
    applicationId: opts.applicationId,
    subject: { type: "agent", userId: opts.userId, agentId: opts.agentId },
  });
  return {
    tools: new ToolGuard(permissions),
    mcp: new MCPGuard(permissions),
    permissions,
  };
}

File: sdks/go/authz/guard.go
package authz

import (
	"context"
	"fmt"
)

// ToolGuard is a local Policy Enforcement Point for tool calls.
//
// Construct it with the precomputed effective permission set; it answers
// per-call permission checks without contacting the AuthZ service.
type ToolGuard struct {
	permissions map[string]struct{}
}

// NewToolGuard creates a guard from a permission set.
func NewToolGuard(permissions map[string]struct{}) *ToolGuard {
	if permissions == nil {
		permissions = map[string]struct{}{}
	}
	return &ToolGuard{permissions: permissions}
}

// IsAllowed returns whether the (resource, action) is in the permission set.
func (g *ToolGuard) IsAllowed(resource, action string) bool {
	_, ok := g.permissions[resource+"."+action]
	return ok
}

// Require returns an error wrapping ErrPermissionDenied when missing.
func (g *ToolGuard) Require(resource, action string) error {
	if !g.IsAllowed(resource, action) {
		return fmt.Errorf("%w: %s.%s", ErrPermissionDenied, resource, action)
	}
	return nil
}

// MCPGuard mirrors ToolGuard but namespaces checks under "mcp.".
type MCPGuard struct {
	permissions map[string]struct{}
}

// NewMCPGuard creates a guard tied to the mcp.* namespace.
func NewMCPGuard(permissions map[string]struct{}) *MCPGuard {
	if permissions == nil {
		permissions = map[string]struct{}{}
	}
	return &MCPGuard{permissions: permissions}
}

// IsAllowed returns whether mcp.<server>.<action> is allowed.
func (g *MCPGuard) IsAllowed(server, action string) bool {
	_, ok := g.permissions["mcp."+server+"."+action]
	return ok
}

// Require returns ErrPermissionDenied when the MCP tool is not allowed.
func (g *MCPGuard) Require(server, action string) error {
	if !g.IsAllowed(server, action) {
		return fmt.Errorf("%w: mcp.%s.%s", ErrPermissionDenied, server, action)
	}
	return nil
}

// StartAgentSession preloads the effective permission set for a (user, agent)
// pair and returns guards bound to that snapshot. Equivalent to the Python
// ``start_agent_session`` helper.
func StartAgentSession(
	ctx context.Context,
	client *Client,
	tenantID, applicationID, userID, agentID string,
) (*ToolGuard, *MCPGuard, error) {
	perms, err := client.GetEffectivePermissions(ctx, EffectivePermissionsRequest{
		TenantID:      tenantID,
		ApplicationID: applicationID,
		Subject: Subject{
			Type:    "agent",
			UserID:  userID,
			AgentID: agentID,
		},
	})
	if err != nil {
		return nil, nil, err
	}
	return NewToolGuard(perms), NewMCPGuard(perms), nil
}
